src/chebyshev_mod.F90

```diff
@@ -0,0 +1,290 @@
+    call get_chebcoeffs(scaledkbt,scaledef,ncoeffs,coeffs,gbnd(1),gbnd(2))
+
+    ! Prepare bml matrices for recursion.
+    call bml_zero_matrix(bml_type,bml_element_real,dp,norb,norb,tnp1_bml)
+    call bml_zero_matrix(bml_type,bml_element_real,dp,norb,norb,tn_bml)
+    call bml_zero_matrix(bml_type,bml_element_real,dp,norb,norb,tnm1_bml)
+    call bml_zero_matrix(bml_type,bml_element_real,dp,norb,norb,aux1_bml)
+
+    ! Set the domain for the tracking function
+    do i=1,N
+       domain0(i) = gbnd(1) + i*de
+       domain0(i) = 2.0_dp*(domain0(i) - gbnd(1))/(gbnd(2)-gbnd(1)) - 1.0_dp
+       tnp1(i) = 0.0_dp
+       tnm1(i) = 1.0_dp
+       tn(i) = domain0(i)
+       domain(i) = 0.0_dp
+    enddo
+
+    !First step of recursion ...
+    mycoeff = jackson(ncoeffs,1)*coeffs(1) !Application of the Jackson kernell
+    call bml_add_identity(tnm1_bml, 1.0_dp, threshold) !T0
+    call bml_scale(mycoeff,tnm1_bml,aux1_bml) !Rho(0) = coeffs(1)*T0
+    domain = 0.0_dp + mycoeff*tnm1
+
+    !Second step of recursion ...
+    mycoeff = jackson(ncoeffs,2)*coeffs(2)
+    call bml_copy(x_bml,tn_bml) !T1
+    call bml_add_deprecated(1.0_dp,aux1_bml,mycoeff,tn_bml,threshold) !Rho(1) = Rho(0) + coeffs(2)*T(1)
+    domain = domain + mycoeff*tn
+
+    !Third to n-1 step of recursion ...
+    if(verbose >= 1) write(*,*)"Chebyshev recursion ..."
+    do i=2,ncoeffs-1
+
+       mycoeff = coeffs(i+1)
+
+       call bml_copy(tnm1_bml,tnp1_bml)
+       tnp1 = tnm1
+       call bml_multiply(x_bml,tn_bml,tnp1_bml,2.0_dp,-1.0_dp) !T(n+1) = 2xT(n) - T(n-1)
+       tnp1 = 2.0_dp*domain0*tn - tnp1
+       mycoeff = mycoeff*jackson(ncoeffs,i)
+
+       call bml_add_deprecated(1.0_dp,aux1_bml,mycoeff,tnp1_bml,threshold) !Rho(n+1) = Rho(n) + b(n+1)*T(n+1)
+       domain = domain + mycoeff*tnp1
+
+       call bml_copy(tn_bml,tnm1_bml)
+       call bml_copy(tnp1_bml,tn_bml)
+       tnm1 = tn
+       tn = tnp1
+
+       occ = 2.0_dp*bml_trace(aux1_bml)
+       if(verbose >= 1) write(*,*)"step, occ",i,occ
+
+    enddo
+
+    if(verbose >= 2) then
+       call open_file(io,"fermi_approx.out")
+       write(io,*)"# Energy, FApprox, Fermi"
+       do i=1,N
+          write(io,*)gbnd(1) + i*de,domain(i),fermi(gbnd(1) + i*de, scaledef, scaledkbt)
+       enddo
+    endif
+
+    if(verbose >= 2) then
+      maxder = absmaxderivative(domain,de)
+      write(*,*)"TargetKbt =",scaledkbt
+      write(*,*)"AbsMaxDerivative =",maxder
+      write(*,*)"kbT = 1/(4*AbsMaxDerivative) =",1.0_dp/(4.0_dp*maxder)
+    endif
+
+    call bml_copy(aux1_bml,rho_bml)
+    call bml_scale(2.0d0, rho_bml)
+
+    write(*,*)"TotalOccupation =", bml_trace(rho_bml)
+
+  end subroutine build_density_cheb
+
+  !> Evaluates the Jackson Kernel Coefficient.
+  !! \param ncoeffs Number of Chebyshev polynomial.
+  !! \param i Coefficient number i.
+  !!
+  real(dp) function jackson(ncoeffs,i)
+    implicit none
+    integer, intent(in) :: ncoeffs,i
+    real(dp) :: pi
+
+    pi = 3.14159265358979323846264338327950_dp
```

Comment:
Since `pi` is used in several subroutines, it should be refactored as a module parameter.